prow/cmd/sinker/main.go

```diff
@@ -532,3 +560,78 @@ func podNeedsKubernetesFinalizerCleanup(log *logrus.Entry, pj *prowapi.ProwJob,
 
 	return false
 }
+
+func getDuplicates(pjs *prowapi.ProwJobList) ([]string, error) {
+	type duplicationKey struct {
+		job string
+		// pointer types are only equal
+		// when they point to the same
+		// address, so we have to serialize
+		// the value of these fields
+		refs      string
+		extraRefs string
+		state     prowapi.ProwJobState
+	}
+	duplicates := map[duplicationKey][]string{}
+
+	var key duplicationKey
+	var err error
+	for _, pj := range pjs.Items {
+		if pj.Spec.Type == prowapi.PeriodicJob || !pj.Complete() {
+			continue
+		}
+
+		key.job = pj.Spec.Job
+		key.refs, err = comparableRefsString(pj.Spec.Refs)
+		if err != nil {
+			return nil, err
+		}
+		key.extraRefs, err = comparableExtraRefsString(pj.Spec.ExtraRefs)
```

Comment:
nit: All this marshalling will probably be pretty expensive. I don't really see a better alternative, but it might be worth logging the time we spend in `getDuplicates`.

Reply:
Yeah I thought a bit about this, but I think we do not care a lot about how long this takes, because in most cases sinker sync takes a lot less time than its resync period. It might be possible to optimize this a bit by handcrafting something to construct a string out of a `ref` but that adds some complexity and I wanted to avoid this unless we are sure the performance is actually an issue.